In [1]:
import pandas as pd
import numpy as np
import json
#import os
import datetime as dt

In [2]:
zillow_home_index_file = "data/Zip_Zhvi_Summary_AllHomes.csv"
zip_code_file = "data/zip_code_list.txt"
zip_code_stats = "data/Zip_Code_Raw_Data.csv"

Note: the realtor_data_file (RDC_InventoryCoreMetrics_Zip_Hist.csv) is too large (250+ MB) to be stored in Github.   
In the following cell, you must specify where to load the file from. 

The first option will obtain it directly from realtor.com.
The second option assumes you downloaded the file and saved it in the data folder.  


In [3]:
# Uncomment the option you want before running the this cell!

#realtor_data = "https://s3-us-west-2.amazonaws.com/econresearch/Reports/Core/RDC_InventoryCoreMetrics_Zip_Hist.csv"
realtor_data = "data/RDC_InventoryCoreMetrics_Zip_Hist.csv"

In [4]:
realtor_df = pd.read_csv(realtor_data, encoding="ISO-8859-1", dtype={'Month':'str', 'ZipCode':'str', 'Footnote':'str'}, skipfooter=1, engine="python")

In [5]:
#realtor_df.info()

In [6]:
#realtor_df['Month'].to_datetime()
realtor_df['Month'] = pd.to_datetime(realtor_df['Month'])

In [7]:
#realtor_df['Month'].dtype

In [8]:
realtor_df.head()

Month ZipCode          ZipName Footnote  Median Listing Price  \
0 2019-02-01   01001       Agawam, MA        *              209500.0   
1 2019-02-01   01002      Amherst, MA      nan              447225.0   
2 2019-02-01   01005        Barre, MA        *              241550.0   
3 2019-02-01   01007  Belchertown, MA      nan              331225.0   
4 2019-02-01   01008    Blandford, MA        *              216275.0   

   Median Listing Price M/M  Median Listing Price Y/Y  Active Listing Count   \
0                    0.0411                       NaN                   47.0   
1                   -0.0150                    0.1540                   56.0   
2                   -0.0069                    0.0679                   16.5   
3                   -0.0076                   -0.0398                   49.0   
4                   -0.0333                       NaN                    9.5   

   Active Listing Count M/M  Active Listing Count Y/Y        ...          \
0                   -0.1532                   -0.1132        ...           
1                    0.1200                   -0.2483        ...           
2                   -0.3529                    0.0313        ...           
3                    0.0889                   -0.2033        ...           
4                    0.0556                       NaN        ...           

   Pending Listing Count Y/Y  Avg Listing Price  Avg Listing Price M/M  \
0                     0.0000           242888.0                 0.0217   
1                        NaN           493027.0                -0.0132   
2                        NaN           343905.0                 0.0343   
3                        NaN           324858.0                -0.0221   
4                     0.3333           224871.0                -0.0243   

   Avg Listing Price Y/Y  Total Listing Count  Total Listing Count M/M  \
0                -0.1314                 48.0                  -0.1504   
1                 0.0617                 56.0                   0.1200   
2                 0.6497                 16.5                  -0.3529   
3                -0.0564                 49.0                   0.0889   
4                 0.0409                 11.5                   0.0455   

   Total Listing Count Y/Y  Pending Ratio  Pending Ratio M/M  \
0                  -0.1111         0.0213             0.0033   
1                  -0.2483         0.0000             0.0000   
2                   0.0313         0.0000             0.0000   
3                  -0.2033         0.0000             0.0000   
4                   0.4375         0.2105            -0.0117   

   Pending Ratio Y/Y  
0             0.0024  
1             0.0000  
2             0.0000  
3             0.0000  
4            -0.0202  

[5 rows x 34 columns]

In [9]:
filtered_realtor_df = realtor_df[realtor_df['Month'].dt.year == 2018]

In [11]:
filtered_realtor_df.head(100)

Month ZipCode               ZipName Footnote  Median Listing Price  \
30057 2018-12-01   01001            Agawam, MA      nan              208300.0   
30058 2018-12-01   01002           Amherst, MA      nan              425050.0   
30059 2018-12-01   01005             Barre, MA        *              240050.0   
30060 2018-12-01   01007       Belchertown, MA      nan              344950.0   
30061 2018-12-01   01008         Blandford, MA        *              239950.0   
30062 2018-12-01   01010         Brimfield, MA      nan              295050.0   
30063 2018-12-01   01013          Chicopee, MA      nan              165050.0   
30064 2018-12-01   01020          Chicopee, MA      nan              181975.0   
30065 2018-12-01   01022          Chicopee, MA        *              151900.0   
30066 2018-12-01   01027       Easthampton, MA      nan              299950.0   
30067 2018-12-01   01028   East Longmeadow, MA      nan              277300.0   
30068 2018-12-01   01030     Feeding Hills, MA      nan              256450.0   
30069 2018-12-01   01033            Granby, MA      nan              324950.0   
30070 2018-12-01   01034         Granville, MA      nan              280050.0   
30071 2018-12-01   01035            Hadley, MA      nan              490000.0   
30072 2018-12-01   01036           Hampden, MA      nan              287450.0   
30073 2018-12-01   01040           Holyoke, MA        *              194950.0   
30074 2018-12-01   01050        Huntington, MA      nan              262550.0   
30075 2018-12-01   01056            Ludlow, MA      nan              262450.0   
30076 2018-12-01   01057            Monson, MA      nan              269450.0   
30077 2018-12-01   01060       Northampton, MA      nan              390950.0   
30078 2018-12-01   01062          Florence, MA      nan              359050.0   
30079 2018-12-01   01069            Palmer, MA        *              215050.0   
30080 2018-12-01   01073       Southampton, MA      nan              339550.0   
30081 2018-12-01   01075      South Hadley, MA      nan              269950.0   
30082 2018-12-01   01077         Southwick, MA      nan              314950.0   
30083 2018-12-01   01082              Ware, MA      nan              237500.0   
30084 2018-12-01   01085         Westfield, MA      nan              280000.0   
30085 2018-12-01   01089  West Springfield, MA        *              249950.0   
30086 2018-12-01   01095         Wilbraham, MA        *              287549.5   
...          ...     ...                   ...      ...                   ...   
30127 2018-12-01   01452       Hubbardston, MA        *              350050.0   
30128 2018-12-01   01453        Leominster, MA      nan              259950.0   
30129 2018-12-01   01460         Littleton, MA        *              520000.0   
30130 2018-12-01   01462         Lunenburg, MA        *              340050.0   
30131 2018-12-01   01463         Pepperell, MA      nan              434950.0   
30132 2018-12-01   01464           Shirley, MA      nan              359050.0   
30133 2018-12-01   01468         Templeton, MA      nan              225000.0   
30134 2018-12-01   01469          Townsend, MA        *              335050.0   
30135 2018-12-01   01473       Westminster, MA      nan              429950.0   
30136 2018-12-01   01475        Winchendon, MA      nan              219950.0   
30137 2018-12-01   01501            Auburn, MA      nan              284950.0   
30138 2018-12-01   01504        Blackstone, MA        *              324550.0   
30139 2018-12-01   01505          Boylston, MA      nan              742550.0   
30140 2018-12-01   01506        Brookfield, MA      nan              224950.0   
30141 2018-12-01   01507          Charlton, MA      nan              322495.0   
30142 2018-12-01   01510           Clinton, MA      nan              294950.0   
30143 2018-12-01   01516           Douglas, MA      nan              366550.0   
30144 2018-12-01   01518          Fiskdale

In [12]:
zips_df = pd.read_csv(zip_code_file, encoding="ISO-8859-1", dtype={'ZipCode':'str'})

In [13]:
#zips_df.info()

In [14]:
# Get a count of zip codes in my zip list
zips_count = zips_df['ZipCode'].count()
zips_count

84

In [31]:
filtered_zips_df = pd.merge(realtor_df, zips_df, on="ZipCode")

In [34]:
#filtered_zips_df.info()

In [33]:
filtered_zips_df

Month ZipCode                 ZipName Footnote  \
0    2019-02-01   92241  Desert Hot Springs, CA      nan   
1    2019-01-01   92241  Desert Hot Springs, CA      nan   
2    2018-12-01   92241  Desert Hot Springs, CA      nan   
3    2018-11-01   92241  Desert Hot Springs, CA      nan   
4    2018-10-01   92241  Desert Hot Springs, CA        *   
5    2018-09-01   92241  Desert Hot Springs, CA      nan   
6    2018-08-01   92241  Desert Hot Springs, CA      nan   
7    2018-07-01   92241  Desert Hot Springs, CA      nan   
8    2018-06-01   92241  Desert Hot Springs, CA      nan   
9    2018-05-01   92241  Desert Hot Springs, CA      nan   
10   2018-04-01   92241  Desert Hot Springs, CA      nan   
11   2018-03-01   92241  Desert Hot Springs, CA      nan   
12   2018-02-01   92241  Desert Hot Springs, CA      nan   
13   2018-01-01   92241  Desert Hot Springs, CA      nan   
14   2017-12-01   92241  Desert Hot Springs, CA      nan   
15   2017-11-01   92241  Desert Hot Springs, CA      nan   
16   2017-10-01   92241  Desert Hot Springs, CA      nan   
17   2017-09-01   92241  Desert Hot Springs, CA      nan   
18   2017-08-01   92241  Desert Hot Springs, CA      nan   
19   2017-07-01   92241  Desert Hot Springs, CA        *   
20   2017-06-01   92241  Desert Hot Springs, CA        *   
21   2017-05-01   92241  Desert Hot Springs, CA      nan   
22   2017-04-01   92241  Desert Hot Springs, CA      nan   
23   2017-03-01   92241  Desert Hot Springs, CA      nan   
24   2017-02-01   92241  Desert Hot Springs, CA      nan   
25   2017-01-01   92241  Desert Hot Springs, CA      nan   
26   2016-12-01   92241  Desert Hot Springs, CA      nan   
27   2016-11-01   92241  Desert Hot Springs, CA      nan   
28   2016-10-01   92241  Desert Hot Springs, CA      nan   
29   2016-09-01   92241  Desert Hot Springs, CA        *   
...         ...     ...                     ...      ...   
3496 2014-10-01   95589          Whitethorn, CA      nan   
3497 2014-09-01   95589          Whitethorn, CA      nan   
3498 2014-08-01   95589          Whitethorn, CA      nan   
3499 2014-07-01   95589          Whitethorn, CA      nan   
3500 2014-06-01   95589          Whitethorn, CA      nan   
3501 2014-05-01   95589          Whitethorn, CA        *   
3502 2014-04-01   95589          Whitethorn, CA        *   
3503 2014-03-01   95589          Whitethorn, CA      nan   
3504 2014-02-01   95589          Whitethorn, CA        *   
3505 2014-01-01   95589          Whitethorn, CA        *   
3506 2013-12-01   95589          Whitethorn, CA        *   
3507 2013-11-01   95589          Whitethorn, CA        *   
3508 2013-10-01   95589          Whitethorn, CA        *   
3509 2013-09-01   95589          Whitethorn, CA        *   
3510 2013-08-01   95589          Whitethorn, CA      nan   
3511 2013-07-01   95589          Whitethorn, CA      nan   
3512 2013-06-01   95589          Whitethorn, CA      nan   
3513 2013-05-01   95589          Whitethorn, CA        *   
3514 2013-04-01   95589          Whitethorn, CA        *   
3515 2013-03-01   95589          Whitethorn, CA      nan   
3516 2013-02-01   95589          Whitethorn, CA      nan   
3517 2013-01-01   95589          Whitethorn, CA      nan   
3518 2012-12-01   95589          Whitethorn, CA      nan   
3519 2012-11-01   95589          Whitethorn, CA      nan   
3520 2012-10-01   95589          Whitethorn, CA      nan   
3521 2012-09-01   95589          Whitethorn, CA      nan   
3522 2012-08-01   95589          Whitethorn, CA      nan   
3523 2012-07-01   95589          Whitethorn, CA      nan   
3524 2012-06-01   95589          Whitethorn, CA      nan   
3525 2012-05-01   95589          Whitethorn, CA      nan   

      Median Listing Price  Median Listing Price M/M  \
0                249950.00                   -0.0090   
1                252225.00                   -0.0089   
2                254500.00                   -0.0745   
3                275000.00                   -0.0516   
4        

In [35]:
filtered_zipcode_count = len(filtered_zips_df['ZipCode'].unique())
filtered_zipcode_count

39

In [36]:
print('Total SAF Zipcodes: ' + str(zips_count))
print('SAF Zipcodes in dataset: ' + str(filtered_zipcode_count))
print('Missing Zipcodes: ' + str(zips_count - filtered_zipcode_count))

Total SAF Zipcodes: 84
SAF Zipcodes in dataset: 39
Missing Zipcodes: 45


In [37]:
zip_listings = filtered_zips_df[['Month', 'ZipCode', 'ZipName', 'Median Listing Price', 'Avg Listing Price']]

In [50]:
yet_another_dataframe_because_rename_doesnt_rename_in_place = zip_listings.rename(columns = {'Month':'Date'})
yet_another_dataframe_because_rename_doesnt_rename_in_place.head()

Date ZipCode                 ZipName  Median Listing Price  \
0 2019-02-01   92241  Desert Hot Springs, CA              249950.0   
1 2019-01-01   92241  Desert Hot Springs, CA              252225.0   
2 2018-12-01   92241  Desert Hot Springs, CA              254500.0   
3 2018-11-01   92241  Desert Hot Springs, CA              275000.0   
4 2018-10-01   92241  Desert Hot Springs, CA              289950.0   

   Avg Listing Price  
0           433049.0  
1           454677.0  
2           488984.0  
3           471347.0  
4           431134.0

In [52]:
zip_listings = yet_another_dataframe_because_rename_doesnt_rename_in_place
zip_listings.head()

Date ZipCode                 ZipName  Median Listing Price  \
0 2019-02-01   92241  Desert Hot Springs, CA              249950.0   
1 2019-01-01   92241  Desert Hot Springs, CA              252225.0   
2 2018-12-01   92241  Desert Hot Springs, CA              254500.0   
3 2018-11-01   92241  Desert Hot Springs, CA              275000.0   
4 2018-10-01   92241  Desert Hot Springs, CA              289950.0   

   Avg Listing Price  
0           433049.0  
1           454677.0  
2           488984.0  
3           471347.0  
4           431134.0

In [69]:
len(zip_listings)

3526

In [57]:
zillow_df = pd.read_csv(zillow_home_index_file, encoding="ISO-8859-1", parse_dates=['Date'])

In [58]:
new_zillow_df = zillow_df.rename(columns={'RegionName':'ZipCode','Zhvi':'zhvi'})

In [59]:
new_zillow_df['ZipName'] = zillow_df['City'].map(str) + ", " + zillow_df['State']

In [60]:
zillow_listings = new_zillow_df[['Date','ZipCode','zhvi']]

In [61]:
zillow_listings = zillow_listings.astype({"ZipCode": str})

In [62]:
cleaned_zillow_df = pd.merge(zillow_listings, zips_df, on="ZipCode")

In [63]:
len(cleaned_zillow_df)

54

In [64]:
cleaned_zillow_df.head()

Date ZipCode    zhvi
0 2019-02-28   95589  334000
1 2019-02-28   95465  986400
2 2019-02-28   95460  696000
3 2019-02-28   95445  403300
4 2019-02-28   95437  368200

In [65]:
new_merged_df = pd.merge(zip_listings, cleaned_zillow_df,  how='left', left_on=['Date','ZipCode'], right_on = ['Date','ZipCode'])

In [67]:
new_merged_df.head()

Date ZipCode                 ZipName  Median Listing Price  \
0 2019-02-01   92241  Desert Hot Springs, CA              249950.0   
1 2019-01-01   92241  Desert Hot Springs, CA              252225.0   
2 2018-12-01   92241  Desert Hot Springs, CA              254500.0   
3 2018-11-01   92241  Desert Hot Springs, CA              275000.0   
4 2018-10-01   92241  Desert Hot Springs, CA              289950.0   

   Avg Listing Price  zhvi  
0           433049.0   NaN  
1           454677.0   NaN  
2           488984.0   NaN  
3           471347.0   NaN  
4           431134.0   NaN

In [68]:
len(new_merged_df)

3526

In [70]:
zip_stats = pd.read_csv(zip_code_stats, encoding="ISO-8859-1")

In [71]:
zip_stats.head(5)

GEO.id GEO.id2 GEO.display-label GCT_STUB.target-geo-id  \
0            Id     Id2         Geography          Target Geo Id   
1  8500000US900     900         ZCTA3 900           8500000US900   
2  8500000US900     900         ZCTA3 900         8600000US90001   
3  8500000US900     900         ZCTA3 900         8600000US90002   
4  8500000US900     900         ZCTA3 900         8600000US90003   

  GCT_STUB.target-geo-id2 GCT_STUB.display-label GCT_STUB.display-label.1  \
0          Target Geo Id2      Geographical Area        Geographical Area   
1                     900               ZCTA 900                 ZCTA 900   
2                   90001  ZCTA 900 - ZCTA 90001               ZCTA 90001   
3                   90002  ZCTA 900 - ZCTA 90002               ZCTA 90002   
4                   90003  ZCTA 900 - ZCTA 90003               ZCTA 90003   

         HC01            HC02                               HC04  \
0  Population  Housing  units  Area in square miles - Total area   
1     2376389          857386                             220.57   
2       54481           13302                               3.47   
3       44584           11838                               3.04   
4       58187           16071                               3.54   

                                HC05                              HC06  \
0  Area in square miles - Water area  Area in square miles - Land area   
1                               1.17                            219.40   
2                               0.00                              3.47   
3                               0.00                              3.04   
4                               0.00                              3.54   

                                                HC08  \
0  Density per square mile of land area - Population   
1                                            10831.3   
2                                            15688.0   
3                                            14677.4   
4                                            16440.6   

                                                HC09  
0  Density per square mile of land area - Housing...  
1                                             3907.9  
2                                             3830.4  
3                                             3897.2  
4                                             4540.8

In [72]:
header = zip_stats.iloc[0]

In [73]:
new_zip_stats = zip_stats[1:]

In [74]:
new_zip_stats.head()

GEO.id GEO.id2 GEO.display-label GCT_STUB.target-geo-id  \
1  8500000US900     900         ZCTA3 900           8500000US900   
2  8500000US900     900         ZCTA3 900         8600000US90001   
3  8500000US900     900         ZCTA3 900         8600000US90002   
4  8500000US900     900         ZCTA3 900         8600000US90003   
5  8500000US900     900         ZCTA3 900         8600000US90004   

  GCT_STUB.target-geo-id2 GCT_STUB.display-label GCT_STUB.display-label.1  \
1                     900               ZCTA 900                 ZCTA 900   
2                   90001  ZCTA 900 - ZCTA 90001               ZCTA 90001   
3                   90002  ZCTA 900 - ZCTA 90002               ZCTA 90002   
4                   90003  ZCTA 900 - ZCTA 90003               ZCTA 90003   
5                   90004  ZCTA 900 - ZCTA 90004               ZCTA 90004   

      HC01    HC02    HC04  HC05    HC06     HC08    HC09  
1  2376389  857386  220.57  1.17  219.40  10831.3  3907.9  
2    54481   13302    3.47  0.00    3.47  15688.0  3830.4  
3    44584   11838    3.04  0.00    3.04  14677.4  3897.2  
4    58187   16071    3.54  0.00    3.54  16440.6  4540.8  
5    67850   23922    3.06  0.00    3.06  22196.4  7825.8

In [75]:
new_zip_stats.columns = header

In [76]:
#zip_stats[['Id', 'Target Geo Id2', 'Population', 'Housing Units', 'Area in square miles - Total Area', 'Density per square mile of land area - Population', 'Density per square mile of land area - Housing units']]
new_zip_stats.columns

Index(['Id', 'Id2', 'Geography', 'Target Geo Id', 'Target Geo Id2',
       'Geographical Area', 'Geographical Area', 'Population',
       'Housing  units', 'Area in square miles - Total area',
       'Area in square miles - Water area', 'Area in square miles - Land area',
       'Density per square mile of land area - Population',
       'Density per square mile of land area - Housing  units'],
      dtype='object', name=0)

In [118]:
new_new_zip_stats = new_zip_stats[['Id', 'Target Geo Id2', 'Population', 'Housing  units', 'Area in square miles - Total area', 'Density per square mile of land area - Population', 'Density per square mile of land area - Housing  units']]

In [119]:
final_zip_stats = new_new_zip_stats[new_new_zip_stats['Target Geo Id2'].str.match('\d{5}$')]

In [120]:
zip_stats_count = len(new_zip_stats)
final_zip_stats_count = len(final_zip_stats)
print("Zip Data: " + str(zip_stats_count) + " rows")
print("Filtered Zip Data: " + str(final_zip_stats_count) + "rows")
print("Removed " + str((zip_stats_count - final_zip_stats_count)) + " non-zip code rows" )

Zip Data: 1890 rows
Filtered Zip Data: 1752rows
Removed 138 non-zip code rows


In [121]:
second_final_df = final_zip_stats.rename(columns={'Target Geo Id2':'ZipCode'})
final_zip_stats = second_final_df

In [122]:
merged_datasets = pd.merge(new_merged_df, final_zip_stats, how="left", on="ZipCode")

In [125]:
#merged_datasets.info()
merged_datasets.head()

Date ZipCode                 ZipName  Median Listing Price  \
0 2019-02-01   92241  Desert Hot Springs, CA              249950.0   
1 2019-01-01   92241  Desert Hot Springs, CA              252225.0   
2 2018-12-01   92241  Desert Hot Springs, CA              254500.0   
3 2018-11-01   92241  Desert Hot Springs, CA              275000.0   
4 2018-10-01   92241  Desert Hot Springs, CA              289950.0   

   Avg Listing Price  zhvi            Id Population Housing  units  \
0           433049.0   NaN  8500000US922       5512           6043   
1           454677.0   NaN  8500000US922       5512           6043   
2           488984.0   NaN  8500000US922       5512           6043   
3           471347.0   NaN  8500000US922       5512           6043   
4           431134.0   NaN  8500000US922       5512           6043   

  Area in square miles - Total area  \
0                            121.23   
1                            121.23   
2                            121.23   
3                            121.23   
4                            121.23   

  Density per square mile of land area - Population  \
0                                              45.5   
1                                              45.5   
2                                              45.5   
3                                              45.5   
4                                              45.5   

  Density per square mile of land area - Housing  units  
0                                               49.8     
1                                               49.8     
2                                               49.8     
3                                               49.8     
4                                               49.8

Columns:
- Date: Obvious. 
- ZipCode: Five digit US zip code
- Median Listing Price: Median listing price from realtor.com
- Avg Listing Price: Mean listing price from realtor.com
- zhvi: Zillow Home Value Index, median zestimate value from Zillow
- Id2: three-digit zip code prefix (900 = 90001, 90002, 90010, etc)
- Population: US Census, total population for zip code
- Housing Units: US Census, total housing units in zip code
- Arewa